### ISSUE TITLE
When collecting payment for an order before the reader is connected, payment failure is shown before the ready state

### ISSUE THREAD

**Issue by jaclync** [woocommerce/woocommerce-ios#11648]
**Describe the bug**

When collecting payment for an order before the reader is connected, connection failure is shown first after connecting with the reader. Without doing anything, it eventually goes to the ready state. The failure state is on the screen for a few seconds and could be a dropoff point, because I almost thought the connection failed until I saw the connecting state later.

**To Reproduce**

Prerequisite: WooPayments is set up for IPP PdfdoF-D-p2, and a card reader is available for the store country.

- Turn off the card reader
- Launch the app
- Go to the orders tab
- Tap on an order that is eligible for IPP
- Tap `Collect Payment`
- Tap `Card Reader`, then tap to turn on the reader and connect --> notice the "payment failure" state is shown for a few seconds before the connecting/ready state is shown

**Screenshots**

Notice the failure screen at 0:11-0:14.

https://github.com/woocommerce/woocommerce-ios/assets/1945542/ddba459d-4364-45ad-86c2-4055ecbc388f

<img width="1366" alt="Screenshot 2024-01-09 at 4 09 47 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/1872f7d2-ea83-4a50-98c2-df1355126b8c">

**Expected behavior**

If the connection is possible, the failure screen shouldn't be shown prematurely.

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: any
- iOS version: iOS 17
- WooCommerce iOS version: 16.8

**Comment by jaclync** [comment]
From @joshheald's comment peaMlT-nT-p2#comment-850, it looks like the issue was already known https://github.com/woocommerce/woocommerce-ios/issues/11483 and fixed in https://github.com/woocommerce/woocommerce-ios/pull/11519. The issue still persists in the latest trunk because the PR was targeting release 16.7 and hasn't been merged to trunk. Let's close this issue once the merge is done.

I was landing on the same fix as in Josh's PR, and was reading about the related code to make sure there are no side effects in case some completion block needs to be triggered. It was a good learning opportunity anyway.

**Comment by joshheald** [comment]
Sorry this caused you extra work @jaclync 

I thought it was in 16.8, which is going out this week, so wasn't concerned that the fix is still not in trunk... but you're right, it was targetting 16.7, which _~has~ should have been merged to `trunk`_ #11628.

I'm trying to figure out why the bug's still there. I'm worried it's going to regress in 16.8 and 16.9.

_Edit_:
Looks like that merge PR was superseded by #11636, which for some reason doesn't have the changes.

**Comment by joshheald** [comment]
https://github.com/woocommerce/woocommerce-ios/pull/11636#issuecomment-1891910617 ~It's a shame it didn't go in 16.7 but that ship has sailed.~ We decided that it didn't meet the requirements for a hotfix, so I'm going to do a beta fix on 16.9 now.

_Edit:_ It did go out in 16.7, it just regressed in 16.8.

**Comment by joshheald** [comment]
Actually, it did go in 16.7, I've just tested the App Store release. It's only missing from 16.8. If we can fix it before 16.8 is released, it may be worth re-submitting – will discuss with release manager.

**Comment by joshheald** [comment]
We are fixing this with a 16.8 beta fix, which @spencertransier will merge back to 16.9 and trunk.

**Comment by jaclync** [comment]
Fixed in trunk, closing now 🎉
